Audio and Video outputs not working

It seems like bentoml is trying to save the files into a temp directory. I get the same error using either the audio output or video output nodes. The image output node still works fine.

As a poor workaround, I could comment out the check in ComfyUI but it's not the best modification: https://github.com/comfyanonymous/ComfyUI/blob/7ed73d12d13c2c389e0469c46c2db635a7d74278/folder_paths.py#L389 

```
2025-08-21T23:56:17-0400 [INFO] [cli] Starting production HTTP BentoServer from "service:ComfyService" listening on http://localhost:3000 (Press CTRL+C to quit)
2025-08-21T23:56:17-0400 [INFO] [entry_service:ComfyService:1] Attaching to ComfyUI server: localhost:8188
2025-08-21T23:56:17-0400 [INFO] [entry_service:ComfyService:1] Service ComfyService initialized
^[[AExecuting workflow: /var/folders/b9/m7p1rfrn6kn3b7mylnbmh23c0000gn/T/bentoml-request-qvttxl48/workflow_9a727cc12a4ea834.json
                                           0% 0:00:00  got prompt
**** ERROR: Saving image outside the output folder is not allowed.
 full_output_folder: /var/folders/b9/m7p1rfrn6kn3b7mylnbmh23c0000gn/T/bentoml-request-qvttxl48
         output_dir: /Users/steyr/workspace/comfy-ui/comfyui/output
         commonpath: /
!!! Exception during processing !!! **** ERROR: Saving image outside the output folder is not allowed.
 full_output_folder: /var/folders/b9/m7p1rfrn6kn3b7mylnbmh23c0000gn/T/bentoml-request-qvttxl48
         output_dir: /Users/steyr/workspace/comfy-ui/comfyui/output
         commonpath: /
Traceback (most recent call last):
  File "/Users/steyr/workspace/comfy-ui/comfyui/execution.py", line 496, in execute
    output_data, output_ui, has_subgraph, has_pending_tasks = await get_output_data(prompt_id, unique_id, obj, input_data_all, execution_block_cb=execution_block_cb, pre_execute_cb=pre_execute_cb, hidden_inputs=hidden_inputs)
                                                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/steyr/workspace/comfy-ui/comfyui/execution.py", line 315, in get_output_data
    return_values = await _async_map_node_over_list(prompt_id, unique_id, obj, input_data_all, obj.FUNCTION, allow_interrupt=True, execution_block_cb=execution_block_cb, pre_execute_cb=pre_execute_cb, hidden_inputs=hidden_inputs)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/steyr/workspace/comfy-ui/comfyui/execution.py", line 289, in _async_map_node_over_list
    await process_inputs(input_dict, i)
  File "/Users/steyr/workspace/comfy-ui/comfyui/execution.py", line 277, in process_inputs
    result = f(**inputs)
             ^^^^^^^^^^^
  File "/Users/steyr/workspace/comfy-ui/comfyui/comfy_extras/nodes_audio.py", line 216, in save_flac
    return save_audio(self, audio, filename_prefix, format, prompt, extra_pnginfo)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/steyr/workspace/comfy-ui/comfyui/comfy_extras/nodes_audio.py", line 96, in save_audio
    full_output_folder, filename, counter, subfolder, filename_prefix = folder_paths.get_save_image_path(filename_prefix, self.output_dir)
                                                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/steyr/workspace/comfy-ui/comfyui/folder_paths.py", line 395, in get_save_image_path
    raise Exception(err)
Exception: **** ERROR: Saving image outside the output folder is not allowed.
 full_output_folder: /var/folders/b9/m7p1rfrn6kn3b7mylnbmh23c0000gn/T/bentoml-request-qvttxl48
         output_dir: /Users/steyr/workspace/comfy-ui/comfyui/output
         commonpath: /

Prompt executed in 0.03 seconds
Error running workflow
{
  "prompt_id": "e9e42161-5e5e-479b-b65a-38a98a0b971a",
  "node_id": "15",
  "node_type": "CPackOutputAudio",
  "executed": [
    "14"
  ],
  "exception_message": "**** ERROR: Saving image outside the output folder is not allowed.\n full_output_folder: 
/var/folders/b9/m7p1rfrn6kn3b7mylnbmh23c0000gn/T/bentoml-request-qvttxl48\n         output_dir: /Users/steyr/workspace/comfy-ui/comfyui/output\n         
commonpath: /\n",
  "exception_type": "Exception",
  "traceback": [
    "  File \"/Users/steyr/workspace/comfy-ui/comfyui/execution.py\", line 496, in execute\n    output_data, output_ui, has_subgraph, has_pending_tasks = await 
get_output_data(prompt_id, unique_id, obj, input_data_all, execution_block_cb=execution_block_cb, pre_execute_cb=pre_execute_cb, hidden_inputs=hidden_inputs)\n  
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
^^\n",
    "  File \"/Users/steyr/workspace/comfy-ui/comfyui/execution.py\", line 315, in get_output_data\n    return_values = await 
_async_map_node_over_list(prompt_id, unique_id, obj, input_data_all, obj.FUNCTION, allow_interrupt=True, execution_block_cb=execution_block_cb, 
pre_execute_cb=pre_execute_cb, hidden_inputs=hidden_inputs)\n                    
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "  File \"/Users/steyr/workspace/comfy-ui/comfyui/execution.py\", line 289, in _async_map_node_over_list\n    await process_inputs(input_dict, i)\n",
    "  File \"/Users/steyr/workspace/comfy-ui/comfyui/execution.py\", line 277, in process_inputs\n    result = f(**inputs)\n             ^^^^^^^^^^^\n",
    "  File \"/Users/steyr/workspace/comfy-ui/comfyui/comfy_extras/nodes_audio.py\", line 216, in save_flac\n    return save_audio(self, audio, filename_prefix, 
format, prompt, extra_pnginfo)\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "  File \"/Users/steyr/workspace/comfy-ui/comfyui/comfy_extras/nodes_audio.py\", line 96, in save_audio\n    full_output_folder, filename, counter, 
subfolder, filename_prefix = folder_paths.get_save_image_path(filename_prefix, self.output_dir)\n                                                                
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
    "  File \"/Users/steyr/workspace/comfy-ui/comfyui/folder_paths.py\", line 395, in get_save_image_path\n    raise Exception(err)\n"
  ],
  "current_inputs": {
    "filename_prefix": [
      "/var/folders/b9/m7p1rfrn6kn3b7mylnbmh23c0000gn/T/bentoml-request-qvttxl48/9a727cc12a4ea83415_"
    ],
    "audio": [
      "{'waveform': tensor([[[ 0.0000e+00, -6.4347e-13, -4.8955e-13,  ..., -2.5729e-07,\n           1.9893e-09,  8.3099e-08]]]), 'sample_rate': 24000}"
    ],
    "prompt": [
      "{'1': {'inputs': {'file': 'AnimateDiff Output 41.mp4', 'video-preview': ''}, 'class_type': 'LoadVideo', '_meta': {'title': 'Load Video'}, 'id': '1', 
'is_changed': [1755829952.2988176]}, '2': {'inputs': {'value': 'test'}, 'class_type': 'CPackInputString', '_meta': {'title': 'input', 'options': {}}, 'id': '2'},
'14': {'inputs': {'audio': 'ansd-1.mp3', 'audioUI': ''}, 'class_type': 'LoadAudio', '_meta': {'title': 'LoadAudio'}, 'id': '14', 'is_changed': 
['a12b690b0bbc30b1feea99d147aa87fb64ee548fad508e84fe846f64c732323c']}, '15': {'inputs': {'filename_prefix': 
'/var/folders/b9/m7p1rfrn6kn3b7mylnbmh23c0000gn/T/bentoml-request-qvttxl48/9a727cc12a4ea83415_', 'audio': ['14', 0]}, 'class_type': 'CPackOutputAudio', '_meta': 
{'title': 'Audio Output'}, 'id': '15'}}"
    ],
    "extra_pnginfo": [
      null
    ]
  },
  "current_outputs": [
    "2",
    "15",
    "14",
    "1"
  ],
  "timestamp": 1755834986482
}
━━━━━━━━━━━━━━━━━━━━                      50% 0:00:00  
2025-08-21T23:56:26-0400 [ERROR] [entry_service:ComfyService:1] Exception on /generate [POST] (trace=438d8036879c62ab0b93f140784dba35,span=c7fe3d14a2d1c185,sampled=0,service.name=ComfyService)
Traceback (most recent call last):
  File "/Users/steyr/workspace/comfy-ui/venv/lib/python3.11/site-packages/_bentoml_impl/server/app.py", line 604, in api_endpoint_wrapper
    resp = await self.api_endpoint(name, request)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/steyr/workspace/comfy-ui/venv/lib/python3.11/site-packages/_bentoml_impl/server/app.py", line 709, in api_endpoint
    output = await self._to_thread(func, *call_args, **call_kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/steyr/workspace/comfy-ui/venv/lib/python3.11/site-packages/_bentoml_impl/server/app.py", line 550, in _to_thread
    output = await anyio.to_thread.run_sync(func, limiter=self._limiter)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/steyr/workspace/comfy-ui/venv/lib/python3.11/site-packages/anyio/to_thread.py", line 56, in run_sync
    return await get_async_backend().run_sync_in_worker_thread(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/steyr/workspace/comfy-ui/venv/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 2470, in run_sync_in_worker_thread
    return await future
           ^^^^^^^^^^^^
  File "/Users/steyr/workspace/comfy-ui/venv/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 967, in run
    result = context.run(func, *args)
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/steyr/workspace/comfy-ui/venv/lib/python3.11/site-packages/_bentoml_sdk/method.py", line 170, in wrapped
    return self.func(instance, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/private/var/folders/b9/m7p1rfrn6kn3b7mylnbmh23c0000gn/T/comfy-pack-28obmir8-bento/service.py", line 118, in generate
    ret = comfy_pack.run_workflow(
          ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/private/var/folders/b9/m7p1rfrn6kn3b7mylnbmh23c0000gn/T/comfy-pack-28obmir8-bento/comfy_pack/run.py", line 279, in run_workflow
    subprocess.run(command, check=True, env=env)
  File "/Users/steyr/.pyenv/versions/3.11.10/lib/python3.11/subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['comfy', '--skip-prompt', '--workspace', '.', 'run', '--workflow', '/var/folders/b9/m7p1rfrn6kn3b7mylnbmh23c0000gn/T/bentoml-request-qvttxl48/workflow_9a727cc12a4ea834.json', '--port', '8188', '--host', 'localhost', '--timeout', '3600', '--wait']' returned non-zero exit status 1.
2025-08-21T23:56:26-0400 [INFO] [entry_service:ComfyService:1] 127.0.0.1:64550 (scheme=http,method=POST,path=/generate,type=application/json,length=26) (status=500,type=application/json,length=74) 851.174ms (trace=438d8036879c62ab0b93f140784dba35,span=c7fe3d14a2d1c185,sampled=0,service.name=ComfyService)
```

--- a/nodes/nodes.py
+++ b/nodes/nodes.py
@@ -1,5 +1,6 @@
 import glob
 import hashlib
+import io
 import json
 import os
 import shutil
@@ -11,8 +12,6 @@ import folder_paths
 import node_helpers
 import numpy as np
 import torch
-from comfy_extras.nodes_audio import SaveAudio
-from comfy_extras.nodes_video import SaveVideo
 from PIL import Image, ImageOps, ImageSequence, PngImagePlugin
 from PIL.PngImagePlugin import PngInfo
 
@@ -466,24 +465,224 @@ class OutputZip:
         return ()
 
 
-class OutputAudio(SaveAudio):
+class OutputAudio:
     CPACK_NODE = True
     CATEGORY = "ComfyPack/output"
+    RETURN_TYPES = ()
+    FUNCTION = "save_audio"
+    OUTPUT_NODE = True
+
+    def __init__(self):
+        self.output_dir = folder_paths.get_output_directory()
+        self.type = "output"
+        self.prefix_append = ""
+
+    def save_audio(
+        self,
+        audio,
+        filename_prefix,
+        format="flac",
+        prompt=None,
+        extra_pnginfo=None,
+        quality="128k",
+    ):
+        import av
+        import torchaudio
+
+        filename_prefix += self.prefix_append
+        full_output_folder, filename, counter, subfolder, filename_prefix = (
+            get_save_image_path(filename_prefix, self.output_dir)
+        )
+        results = []
+
+        # Prepare metadata dictionary
+        metadata = {}
+        if prompt is not None:
+            metadata["prompt"] = json.dumps(prompt)
+        if extra_pnginfo is not None:
+            for x in extra_pnginfo:
+                metadata[x] = json.dumps(extra_pnginfo[x])
+
+        # Opus supported sample rates
+        OPUS_RATES = [8000, 12000, 16000, 24000, 48000]
+
+        for batch_number, waveform in enumerate(audio["waveform"].cpu()):
+            filename_with_batch_num = filename.replace("%batch_num%", str(batch_number))
+            file = f"{filename_with_batch_num}_{counter:05}.{format}"
+            output_path = os.path.join(full_output_folder, file)
+
+            # Use original sample rate initially
+            sample_rate = audio["sample_rate"]
+
+            # Handle Opus sample rate requirements
+            if format == "opus":
+                if sample_rate > 48000:
+                    sample_rate = 48000
+                elif sample_rate not in OPUS_RATES:
+                    # Find the next highest supported rate
+                    for rate in sorted(OPUS_RATES):
+                        if rate > sample_rate:
+                            sample_rate = rate
+                            break
+                    if sample_rate not in OPUS_RATES:  # Fallback if still not supported
+                        sample_rate = 48000
+
+                # Resample if necessary
+                if sample_rate != audio["sample_rate"]:
+                    waveform = torchaudio.functional.resample(
+                        waveform, audio["sample_rate"], sample_rate
+                    )
+
+            # Create in-memory WAV buffer
+            wav_buffer = io.BytesIO()
+            torchaudio.save(wav_buffer, waveform, sample_rate, format="WAV")
+            wav_buffer.seek(0)  # Rewind for reading
+
+            # Use PyAV to convert and add metadata
+            input_container = av.open(wav_buffer)
+
+            # Create output with specified format
+            output_buffer = io.BytesIO()
+            output_container = av.open(output_buffer, mode="w", format=format)
+
+            # Set metadata on the container
+            for key, value in metadata.items():
+                output_container.metadata[key] = value
+
+            # Set up the output stream with appropriate properties
+            input_container.streams.audio[0]
+            if format == "opus":
+                out_stream = output_container.add_stream("libopus", rate=sample_rate)
+                if quality == "64k":
+                    out_stream.bit_rate = 64000
+                elif quality == "96k":
+                    out_stream.bit_rate = 96000
+                elif quality == "128k":
+                    out_stream.bit_rate = 128000
+                elif quality == "192k":
+                    out_stream.bit_rate = 192000
+                elif quality == "320k":
+                    out_stream.bit_rate = 320000
+            elif format == "mp3":
+                out_stream = output_container.add_stream("libmp3lame", rate=sample_rate)
+                if quality == "V0":
+                    # TODO i would really love to support V3 and V5 but there doesn't seem to be a way to set the qscale level, the property below is a bool
+                    out_stream.codec_context.qscale = 1
+                elif quality == "128k":
+                    out_stream.bit_rate = 128000
+                elif quality == "320k":
+                    out_stream.bit_rate = 320000
+            else:  # format == "flac":
+                out_stream = output_container.add_stream("flac", rate=sample_rate)
+
+            # Copy frames from input to output
+            for frame in input_container.decode(audio=0):
+                frame.pts = None  # Let PyAV handle timestamps
+                output_container.mux(out_stream.encode(frame))
+
+            # Flush encoder
+            output_container.mux(out_stream.encode(None))
+
+            # Close containers
+            output_container.close()
+            input_container.close()
+
+            # Write the output to file
+            output_buffer.seek(0)
+            with open(output_path, "wb") as f:
+                f.write(output_buffer.getbuffer())
+
+            results.append(
+                {"filename": file, "subfolder": subfolder, "type": self.type}
+            )
+            counter += 1
+
+        return {"ui": {"audio": results}}
 
     @classmethod
     def INPUT_TYPES(s):
         return {
             "required": {
                 "audio": ("AUDIO",),
-                "filename_prefix": ("STRING", {"default": "cpack_output_"}),
+                "filename_prefix": ("STRING", {"default": "audio/comfypack"}),
             },
             "hidden": {"prompt": "PROMPT", "extra_pnginfo": "EXTRA_PNGINFO"},
         }
 
 
-class OutputVideo(SaveVideo):
+class OutputVideo:
     CPACK_NODE = True
     CATEGORY = "ComfyPack/output"
+
+    RETURN_TYPES = ()
+    FUNCTION = "save_video"
+    DESCRIPTION = "Saves the input video to your ComfyUI output directory."
+    OUTPUT_NODE = True
+
+    def __init__(self):
+        self.output_dir = folder_paths.get_output_directory()
+        self.type = "output"
+        self.prefix_append = ""
+
+    @classmethod
+    def INPUT_TYPES(cls):
+        return {
+            "required": {
+                "video": ("VIDEO", {"tooltip": "The video to save."}),
+                "filename_prefix": (
+                    "STRING",
+                    {
+                        "default": "video/comfypack",
+                        "tooltip": "The prefix for the file to save. This may include formatting information such as %date:yyyy-MM-dd% or %Empty Latent Image.width% to include values from nodes.",
+                    },
+                ),
+                "format": (
+                    ["auto", "mp4"],
+                    {"default": "auto", "tooltip": "The format to save the video as."},
+                ),
+                "codec": (
+                    ["auto", "h264"],
+                    {"default": "auto", "tooltip": "The codec to use for the video."},
+                ),
+            },
+            "hidden": {"prompt": "PROMPT", "extra_pnginfo": "EXTRA_PNGINFO"},
+        }
+
+    def save_video(
+        self,
+        video,
+        filename_prefix,
+        format,
+        codec,
+        prompt=None,
+        extra_pnginfo=None,
+    ):
+        filename_prefix += self.prefix_append
+        width, height = video.get_dimensions()
+        full_output_folder, filename, counter, subfolder, filename_prefix = (
+            get_save_image_path(filename_prefix, self.output_dir, width, height)
+        )
+        results = list()
+        saved_metadata = None
+        metadata = {}
+        if extra_pnginfo is not None:
+            metadata.update(extra_pnginfo)
+        if prompt is not None:
+            metadata["prompt"] = prompt
+        if len(metadata) > 0:
+            saved_metadata = metadata
+        file = f"{filename}_{counter:05}.mp4"
+        video.save_to(
+            os.path.join(full_output_folder, file),
+            format=format,
+            codec=codec,
+            metadata=saved_metadata,
+        )
+
+        results.append({"filename": file, "subfolder": subfolder, "type": self.type})
+        counter += 1
+
+        return {"ui": {"images": results, "animated": (True,)}}
 
 
 class OutputTextFile:

